### PR TITLE
INSP: Add error for `repr` without parentheses

### LINKED ISSUE #6336: Show error on `repr` attribute without parenthesis
## Environment

* **IntelliJ Rust plugin version:** 0.3.135.3449-202-nightly
* **Rust toolchain version:** 1.49.0-nightly (ffa2e7ae8 2020-10-24) x86_64-apple-darwin
* **IDE name and version:** IntelliJ IDEA 2020.2.3 Ultimate Edition (IU-202.7660.26)
* **Operating system:** macOS 10.15.6

## Problem description
Currently, the plugin doesn't produce error annotation on `repr` attribute without parenthesis although `rustc` produces "error: malformed `repr` attribute input" error

image

## Steps to reproduce
```rust
#[repr]
pub enum Color {
    Black = 0
}

```

### PR DESCRIPTION
**Pull request by zeroeightysix**
Adds a new error for when a `repr` attribute does not have `RsMetaItemArgs` (thus, no pair of parentheses)

image

Fixes #6336

changelog: Show error if `repr` attribute doesn't have parentheses and provide quick-fix to add them

### CODE REVIEW

**Comment by zeroeightysix** [comment]
Draft because I'm not sure whether or not to add a `Add parentheses` fix in this PR, or another one. What would you prefer?

**Comment by Undin** [comment]
@zeroeightysix I think `Add parentheses` quick fix is quite simple so you can put it into this PR

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt [hunk 1]
```diff
@@ -1284,6 +1284,14 @@ sealed class RsDiagnostic(
         )
     }
 
+    class ReprNoParentheses(element: PsiElement) : RsDiagnostic(element) {
+        override fun prepare(): PreparedAnnotation = PreparedAnnotation(
+            ERROR,
+            null,
+            "Malformed `repr` attribute input: missing parentheses"
```
Comment (Stzx) [inline src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt:1291]:
```suggestion
            "Malformed `repr` attribute input: missing parentheses",
            fixes = listOfNotNull(AddReprParenthesesFix(element))
```

**Review by Stzx — CHANGES_REQUESTED** [review]

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt (hunk 1 above)
Comment (Stzx) [reply src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt:1291]: Hey, don't forget to add `AddReprParenthesesFix` to the fix list.

Comment (zeroeightysix) [reply src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt]: I had to reread this a few times - I was sure I did this locally, and I did. No clue why I didn't commit it.

File: src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt [hunk 2]
```diff
@@ -909,7 +909,13 @@ class RsErrorAnnotator : AnnotatorBase(), HighlightRangeExtension {
 
         val owner = attr.owner ?: return
 
-        val reprArgs = attr.metaItem.metaItemArgs?.metaItemList.orEmpty()
+        val metaItemArgs = attr.metaItem.metaItemArgs
+
+        if (metaItemArgs == null) {
+            RsDiagnostic.ReprNoParentheses(attr).addToHolder(holder)
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/RsErrorAnnotator.kt:915]: Let's annotate only the corresponding `attr.metaItem` to specify the error range more preciesely.
Also, it allows us to correctly use the same diagnostic for cases like `#[cfg_attr(unix, repr)]` in future to highlight only `repr` text

File: src/main/kotlin/org/rust/ide/annotator/fixes/AddReprParenthesesFix.kt [hunk 3]
```diff
@@ -0,0 +1,31 @@
...
+import com.intellij.codeInspection.LocalQuickFixAndIntentionActionOnPsiElement
+import com.intellij.ide.util.PsiNavigationSupport
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.RsOuterAttr
+import org.rust.lang.core.psi.RsPsiFactory
+
+class AddReprParenthesesFix(element: PsiElement) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/fixes/AddReprParenthesesFix.kt:17]: Optional: to make this quick fox more generic, I suggest to name it `AddParenthesesFix` or `AddAttrParenthesesFix`

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt [hunk 4]
```diff
@@ -1284,6 +1284,15 @@ sealed class RsDiagnostic(
...
     }
 
+    class ReprNoParentheses(element: PsiElement) : RsDiagnostic(element) {
+        override fun prepare(): PreparedAnnotation = PreparedAnnotation(
+            ERROR,
+            null,
+            "Malformed `repr` attribute input: missing parentheses",
+            fixes = listOf(AddReprParenthesesFix(element))
+        )
+    }
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt:1294]: Similar comments as for `AddReprParenthesesFix`:
- let's specify argument type more strictly
- optional: let's rename it to have more generic diagnostic class, for example, `NoAttrParentheses`
- optional: let's retrieve name from the corresponding meta items for error text

**Review by Undin — COMMENTED** [review]
In general looks ok except some comments.

Since this diagnostic can be quite easily applied for different attributes, I've added some optional comments about generalization. If you desired to improve it, you may fix these comments in separate PR with supporting more attributes

File: src/main/kotlin/org/rust/ide/annotator/fixes/AddAttrParenthesesFix.kt [hunk 5]
```diff
@@ -0,0 +1,31 @@
...
+import com.intellij.codeInspection.LocalQuickFixAndIntentionActionOnPsiElement
+import com.intellij.ide.util.PsiNavigationSupport
+import com.intellij.openapi.editor.Editor
+import com.intellij.openapi.project.Project
+import com.intellij.psi.PsiElement
+import com.intellij.psi.PsiFile
+import org.rust.lang.core.psi.RsMetaItem
+import org.rust.lang.core.psi.RsPsiFactory
+
+class AddAttrParenthesesFix(element: RsMetaItem, private val metaItemName: String) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/fixes/AddAttrParenthesesFix.kt:17]:
```suggestion
class AddAttrParenthesesFix(element: RsMetaItem, private val attrName: String) : LocalQuickFixAndIntentionActionOnPsiElement(element) {
```

File: src/main/kotlin/org/rust/ide/annotator/fixes/AddAttrParenthesesFix.kt [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+    override fun getFamilyName(): String = "Add parentheses"
+    override fun getText(): String = "Add parentheses to `$metaItemName`"
+
+    override fun invoke(project: Project, file: PsiFile, editor: Editor?, startElement: PsiElement, endElement: PsiElement) {
+        if (startElement !is RsMetaItem) return
+
+        val newItem = RsPsiFactory(project).createOuterAttr("repr()").metaItem
```
Comment (Undin) [inline src/main/kotlin/org/rust/ide/annotator/fixes/AddAttrParenthesesFix.kt:24]:
```suggestion
        val newItem = RsPsiFactory(project).createOuterAttr("$metaItemName()").metaItem
```

File: src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt [hunk 7]
```diff
@@ -1284,6 +1284,15 @@ sealed class RsDiagnostic(
         )
     }
 
+    class NoAttrParentheses(element: RsMetaItem, private val metaItemName: String) : RsDiagnostic(element) {
```
Comment (Undin) [inline src/main/kotlin/org/rust/lang/utils/RsDiagnostic.kt:1287]:
```suggestion
    class NoAttrParentheses(element: RsMetaItem, private val attrName: String) : RsDiagnostic(element) {
```

**Review by Undin — COMMENTED** [review]
LGTM except comments

Could you also squash all "fixing" commits into main ones, please?

**Review by Undin — APPROVED** [review]
Great!
Thank you!

bors r+

**Comment by Undin** [comment]
@zeroeightysix Would you like to extend this diagnostic for other standard attributes like `allow`, `derive`, etc?

**Comment by zeroeightysix** [comment]
> @zeroeightysix Would you like to extend this diagnostic for other standard attributes like `allow`, `derive`, etc?

Surely, but in another PR if it's okay with you.

**Comment by Undin** [comment]
@zeroeightysix 👍 

> Surely, but in another PR if it's okay with you.

Yep, of course. `bors` is already making final checks to merge this PR into master

**Pull request merged**